x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java

```diff
@@ -35,61 +36,173 @@
+                config.getPackagedModelId() + ModelLoaderUtils.MODEL_FILE_EXTENSION
+            );
+            modelInputStream = ModelLoaderUtils.getInputStreamFromModelRepository(uri);
+
+            if (config.getVocabularyFile() != null) {
+                vocabularyParts = ModelLoaderUtils.loadVocabulary(
+                    ModelLoaderUtils.resolvePackageLocation(config.getModelRepository(), config.getVocabularyFile())
+                );
+            }
+        } catch (Exception e) {
+            finalListener.onFailure(e);
+            return;
         }
 
-        URI uri = ModelLoaderUtils.resolvePackageLocation(
-            config.getModelRepository(),
-            config.getPackagedModelId() + ModelLoaderUtils.MODEL_FILE_EXTENSION
+        downloadParts(
+            new ModelLoaderUtils.InputStreamChunker(modelInputStream, DEFAULT_CHUNK_SIZE, totalParts),
+            size,
+            vocabularyParts,
+            finalListener
         );
+    }
 
-        InputStream modelInputStream = ModelLoaderUtils.getInputStreamFromModelRepository(uri);
-
-        ModelLoaderUtils.InputStreamChunker chunkIterator = new ModelLoaderUtils.InputStreamChunker(modelInputStream, DEFAULT_CHUNK_SIZE);
+    void downloadParts(
+        ModelLoaderUtils.InputStreamChunker chunkIterator,
+        long size,
+        @Nullable ModelLoaderUtils.VocabularyParts vocabularyParts,
+        ActionListener<AcknowledgedResponse> finalListener
+    ) {
+        var countingListener = new RefCountingListener(1, finalListener.map(ignored -> {
+            checkDownloadComplete(chunkIterator);
+            return AcknowledgedResponse.TRUE;
+        }));
+        try {
+            // Uploading other artefacts of the model first, that way the model is last and a simple search can be used to check if the
+            // download is complete
+            if (vocabularyParts != null) {
+                uploadVocabulary(vocabularyParts, countingListener.acquire(r -> {
+                    logger.debug(() -> format("[%s] imported model vocabulary [%s]", modelId, config.getVocabularyFile()));
+                }));
+            }
+
+            for (int part = 0; part < MAX_IN_FLIGHT_REQUESTS; ++part) {
+                if (countingListener.isFailing()) {
+                    break;
+                }
+
+                task.setProgress(chunkIterator.getTotalParts(), chunkIterator.getCurrentPart().get());
+                BytesArray definition = chunkIterator.next();
+
+                if (task.isCancelled()) {
+                    throw new TaskCancelledException(format("task cancelled with reason [%s]", task.getReasonCancelled()));
+                }
+
+                uploadPart(
+                    part,
+                    chunkIterator.getTotalParts(),
+                    size,
+                    definition,
+                    countingListener.acquire(r -> executorService.execute(() -> doNextPart(size, chunkIterator, countingListener)))
```

Comment:
That's lock free because we rely on the thread pool having only one thread? 
Isn't it equivalent to what we have now where a single thread does the download+upload?

Reply:
> That's lock free because we rely on the thread pool having only one thread? 

Yes, it only works because the single thread executor limits concurrent read access to the input stream.

> Isn't it equivalent to what we have now where a single thread does the download+upload?

The upload (index part) is now async which is where we get the speed up and the indexing occurs on a different thread.